libkbfs/folder_branch_ops.go

```diff
@@ -1024,6 +1037,296 @@ func (fbo *folderBranchOps) kickOffPartialSyncIfNeeded(
+	}
+
+	md, err := fbo.getLatestMergedMD(ctx, lState)
+	if err != nil {
+		fbo.log.CDebugf(ctx, "Couldn't get latest merged MD: %+v", err)
+		return nil, nil, 0, nil
+	}
+	if md == (ImmutableRootMetadata{}) ||
+		md.Revision() == kbfsmd.RevisionUninitialized {
+		return nil, nil, 0, errors.New("Unexpectedly no merged revision")
+	}
+
+	// Skip mark-and-sweep if we were woken up by the timer and
+	// the revision hasn't changed since last time.
+	if !triggered && md.Revision() == lastMDRev {
+		fbo.log.CDebugf(
+			ctx, "Revision hasn't changed since last mark-and-sweep")
+		return nil, nil, 0, nil
+	}
+
+	syncConfig, err := fbo.getProtocolSyncConfigUnlocked(ctx, lState, md)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	if syncConfig.Mode != keybase1.FolderSyncMode_PARTIAL {
+		return nil, nil, 0, errors.New("No partial sync config")
+	}
+
+	// Kick off the mark-and-sweep, and wait for it to finish or
+	// be pre-empted.
+	currMarkAndSweepCtxDone, currMarkAndSweepCancel :=
+		fbo.kickOffPartialMarkAndSweep(ctx, lState, syncConfig, md)
+	return currMarkAndSweepCtxDone, currMarkAndSweepCancel, md.Revision(), nil
+}
+
+func (fbo *folderBranchOps) partialMarkAndSweepLoop(trigger <-chan struct{}) {
+	// For partially-synced TLFs, run this:
+	// * Once an hour-ish, only if the latest merged revision has changed.
+	// * When a path is removed from the config.
+	//
+	// Cancel the running mark-and-sweep when:
+	// * The latest merged revision changes.
+	// * The config changes.
+	//
+	// Exit this loop:
+	// * On shutdown.
+	// * When no longer configured to be partially syncing.
+	ctx, cancel := context.WithCancel(fbo.ctxWithFBOID(context.Background()))
+	defer cancel()
+	fbo.log.CDebugf(ctx, "Starting mark-and-sweep loop")
+
+	// Set the first timer to be some random duration less than the
+	// period, to spread out the work of different TLFs.
+	d := time.Duration(rand.Int63n(int64(markAndSweepPeriod)))
+	timer := time.NewTimer(d)
+
+	var currMarkAndSweepCtxDone <-chan struct{}
+	var currMarkAndSweepCancel context.CancelFunc
+	defer func() {
+		if currMarkAndSweepCancel != nil {
+			currMarkAndSweepCancel()
+		}
+		timer.Stop()
+	}()
+
+	lastMDRev := kbfsmd.RevisionUninitialized
+
+	fbo.log.CDebugf(ctx, "Scheduling first timer for %s", d)
+	lState := makeFBOLockState()
+	for {
+		triggered := false
+		select {
+		case <-currMarkAndSweepCtxDone:
+			fbo.log.CDebugf(ctx, "Mark-and-sweep finished; resetting timer")
+			timer = time.NewTimer(markAndSweepPeriod)
+			currMarkAndSweepCtxDone = nil
+			continue
+		case _, ok := <-trigger:
+			if !ok {
+				fbo.log.CDebugf(ctx, "Mark-and-sweep is shutting down.")
+				return
+			}
+			fbo.log.CDebugf(ctx, "New mark-and-sweep triggered")
+			triggered = true
+		case <-timer.C:
+			fbo.log.CDebugf(ctx, "Mark-and-sweep timer fired")
+		case <-fbo.shutdownChan:
+			fbo.log.CDebugf(ctx, "Shutdown")
+			return
+		}
+
+		if currMarkAndSweepCancel != nil {
+			currMarkAndSweepCancel()
+		}
+		timer.Stop()
```

Comment:
why stop the timer here?

Reply:
That's for the case where we got a manual trigger.  No need to let the timer keep going in that case.